### PR TITLE
Implements function movingWindow and child functions

### PR DESCRIPTION
**Pull request by vaguiar**
**Describe your changes**
This PR implements the movingWindow and it's children functions like movingAverage etc. natively. Note: There's no support for fetch a time slice based on number of points

**Testing performed**

- A/B testing with the [mtcmptest](https://github.com/shanson7/mtcmptest) tool.
- Unit tests
- Integration tests

**Additional context**

- Wrote a common function to do xFilesFactor checks for a series of points which is used here and in `removeEmptySeries`
- Added a common util to get a named Series list/array with from/to time specified - `getTimeRangeSeriesListNamed`
- Removed old unstable implementation of `func_movingaverage`

### CODE REVIEW

**Review by robert-milan — APPROVED** [review]

**Comment by robert-milan** [comment]
Just one thing before merging, could you update `docs/graphite.md`? I can in a separate PR also.

File: expr/func_removeemptyseries.go
Comment (Dieterbe) [inline expr/func_removeemptyseries.go:51]: why count all the notnull's and not use it? probably should remove that code

Comment (Dieterbe) [inline expr/func_removeemptyseries.go:51]: why count all the notnull's and not use it? probably should remove that code

File: expr/func_movingwindow.go
Comment (Dieterbe) [inline expr/func_movingwindow.go:43]: the signature is correct for movingWindow but not the other ones where the func is already known

File: expr/func_movingwindow.go
Comment (Dieterbe) [inline expr/func_movingwindow.go:157]: no need to account for very unlikely future changes

File: expr/func_movingwindow.go
Comment (Dieterbe) [inline expr/func_movingwindow.go:41]: incompatibilities with graphite should be documented in docs/graphite.md (@robert-milan please keep this in mind)

File: expr/func_movingwindow.go
Comment (Dieterbe) [inline expr/func_movingwindow.go:99]: windowsize, offset, interval. way too many terms for the same thing.

File: expr/func_movingwindow.go
Comment (Dieterbe) [inline expr/func_movingwindow.go:107]: huh in graphite you can request from=X to=X+12hr and request a windowsize of 24hr for example
it returns a series. why do we return an error?

Comment (Dieterbe) [reply expr/func_movingwindow.go:107]: maybe i'm missing something, because testing it in grafana works fine

File: expr/func_movingaverage.go
Comment (Dieterbe) [inline expr/func_movingaverage.go:25]: this problem seems to have been solved thanks to the context from adjustment.
e.g. with 6h retention, 3h of recent data + 3h window fetches raw, but 4h window goes to the first rollup

**Review by Dieterbe — CHANGES_REQUESTED** [review]
seems to work well for the most part, but needs some tweaks

File: expr/func_movingwindow.go
Comment (vaguiar) [reply expr/func_movingwindow.go:157]: ok. I'll take this out

File: expr/func_movingwindow.go
Comment (vaguiar) [reply expr/func_movingwindow.go:43]: No sure I understand the comment here. For the children of movingWindow, we should use the `NewMovingWindowConstructor` constructor with the func passed in.

File: expr/func_movingwindow.go
Comment (vaguiar) [reply expr/func_movingwindow.go:99]: Agreed that `offset` and `interval` are the same thing and refer to the windowSize in seconds. I can change this.
`windowSize` is the string argument passed in.

File: expr/func_movingwindow.go
Comment (vaguiar) [reply expr/func_movingwindow.go:107]: In your example above `queryStart` is X and `serieStart` is 'X-24hrs' and `serieEnd` is X+12hrs.
This sanity check is to make sure that `querystart` is never calculated to be before `serieStart` and after `serieEnd`. However unlikely it is to occur, we want to check to make sure we error out in that case.

File: expr/func_removeemptyseries.go
Comment (vaguiar) [reply expr/func_removeemptyseries.go:51]: Good catch. I'll take this out

File: expr/func_movingwindow.go
Comment (vaguiar) [reply expr/func_movingwindow.go:41]: Should I mark it `stable` in docs/graphite.md? Just clarifying.

File: expr/func_movingwindow.go
Comment (Dieterbe) [reply expr/func_movingwindow.go:43]: yes but the Signature() function is what is used to validate the query (function arguments) provided by the user.
if we already set the func at construction time, we shouldn't look for an input parameter for it.
See NewAggregate() / FuncAggregate.Signature(), it's the same idea.

File: expr/func_movingwindow.go
Comment (Dieterbe) [reply expr/func_movingwindow.go:41]: so, the fact that we only allow a window size as interval string, not as a integer number of points, is an exception to document.
if, that aside, everything else behaves exactly like graphite does - to your knowledge - then yes you can mark it as stable.

File: expr/func_movingwindow.go
Comment (Dieterbe) [reply expr/func_movingwindow.go:107]: I see. well given how we adjust context.from in FuncMovingWindow.Context() (we reduce `from`), serieStart will always be before querystart.  but ok, i have no objection against making sure.I do think these lines of code need some clarification about why they're checking this, what the concern is, how things might break if the condition is not met, etc

Comment (Dieterbe) [reply expr/func_movingwindow.go:107]: could be as simple as "querystart needs to be between serieStart and serieEnd for the loop below to work (this should always be the case)" . at least that's my interpretation. maybe you can explain it better.

Comment (vaguiar) [reply expr/func_movingwindow.go:107]: Understood. I'll add a comment explaining.

**Comment by vaguiar** [comment]
@Dieterbe & @robert-milan, thanks for your feedback and I've made the requested changes.

**Review by robert-milan — APPROVED** [review]

**Comment by vaguiar** [comment]
Sean noticed an interesting case today where the `movingWindow` function itself gets proxied out to grafana-web, but all its children like `movingAverage` etc, don't. I don't remember seeing this behavior with our previous internal (firmwide) release of MT or with my integration testing. Not sure if this came about during the new changes to the Signature functions.

I'll spend some time looking into and we should hold off on a merge until then.

**Comment by Dieterbe** [comment]
strange, according to `"movingWindow":                 {NewMovingWindow, true}` it should work.
the other reason why it might proxy is if you have this
```
# proxy to graphite when metrictank considers the request bad
proxy-bad-requests = true
```
AND metrictank considers the query invalid (e.g. bad parameters not matching signature). the logs should say

**Comment by Dieterbe** [comment]
@vaguiar any news on this? robert and i think the code looks good, and I'ld like to merge this PR.
if you need assintance, let us know.

**Comment by vaguiar** [comment]
@Dieterbe - After wrapping up some other items, I've picked this up now. I'll try to revert back very soon. Thanks!

**Comment by Dieterbe** [comment]
hi @vaguiar good to hear. what do you mean with "revert"?  as a reminder, robert and I think this code is fine and can be merged. you said there seemed to be an issue?

**Comment by vaguiar** [comment]
> hi @vaguiar good to hear. what do you mean with "revert"? as a reminder, robert and I think this code is fine and can be merged. you said there seemed to be an issue?

By revert back, I meant I'll get back to you :) I'm testing this locally now.

**Comment by vaguiar** [comment]
Hi @Dieterbe - I had my [func signature ](https://github.com/grafana/metrictank/pull/1874/files#diff-15e2ae91f876a86d61765f8ccebb0f8b2d2f7c7889cad01c5006e1a4c93dc994R53) for the base call to `movingWindow` missing the `fn` argument, which I didn't think I needed. 

However, that was giving me the `Proxying to Graphite because of error: argument bad type. expected float - got etString` message and causing it to proxy away from the native implementation.

I've fixed that and rebased off master.

**Comment by Dieterbe** [comment]
your previous version was correct. 
For constructions of known methods such as movingmax etc where you already explicitly set the func...
```
+               "movingAverage":                {NewMovingWindowConstructor("average"), true},
+               "movingMax":                    {NewMovingWindowConstructor("max"), true},
+               "movingMedian":                 {NewMovingWindowConstructor("median"), true},
+               "movingMin":                    {NewMovingWindowConstructor("min"), true},
+               "movingSum":                    {NewMovingWindowConstructor("sum"), true},
+               "movingWindow":                 {NewMovingWindow, true},
```
there is no need to declare "func" as an argument for them.
in fact if you look at the [docs](https://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.movingMax) you can see these functions have no "func" paramether, nor should they, as the function is already known.

what exact query are you using? are you specifying a func in a call that doesn't expect a func argument?

**Comment by Dieterbe** [comment]
also, you should update testMovingWindow to be in line with similar test helpers for other functions. to test whether the input is not changed, whether the output is canonical, DoesNotDoubleReturnPoints, etc. for an example see `testAggregate()`

**Comment by vaguiar** [comment]
> your previous version was correct.
> For constructions of known methods such as movingmax etc where you already explicitly set the func, like ...
> 
> ```
> +               "movingAverage":                {NewMovingWindowConstructor("average"), true},
> +               "movingMax":                    {NewMovingWindowConstructor("max"), true},
> +               "movingMedian":                 {NewMovingWindowConstructor("median"), true},
> +               "movingMin":                    {NewMovingWindowConstructor("min"), true},
> +               "movingSum":                    {NewMovingWindowConstructor("sum"), true},
> ```
> 
> there is no need to declare "func" as an argument for them.
> in fact if you look at the [docs](https://graphite.readthedocs.io/en/latest/functions.html#graphite.render.functions.movingMax) you can see these functions have no "func" paramether, nor should they, as the function is already known.
> 
> what exact query are you using? are you specifying a func in a call that doesn't expect a func argument?

When I use the query builder on Grafana, it sets the func param to `average`. Based on the traces I printed, the order in which it gets executed is:
```
metrictank-query_1  | In NewMovingWindow Constructor
metrictank-query_1  | In Signature Func
metrictank-query_1  | 2020-11-06 15:46:27.310 [INFO] Proxying to Graphite because of error: argument bad type. expected float - got etString
```

Point is if the NewMovingWindow Constructor is defined as [such](https://github.com/grafana/metrictank/pull/1874/files#diff-15e2ae91f876a86d61765f8ccebb0f8b2d2f7c7889cad01c5006e1a4c93dc994R34-R36), wouldn't we always have the fn arg set, thus returning [this](https://github.com/grafana/metrictank/pull/1874/files#diff-15e2ae91f876a86d61765f8ccebb0f8b2d2f7c7889cad01c5006e1a4c93dc994R50-R58) Arg list?

Aggregate does it [differently](https://github.com/grafana/metrictank/blob/master/expr/func_aggregate.go#L24-L41) without setting a default

**Comment by Dieterbe** [comment]
I see. the difference is that `aggregate()` doesn't set a default `func` whereas `movingWindow` does (see https://graphite.readthedocs.io/en/latest/functions.html)
so we can't use the `fn` parameter to differentiate whether the invocation was "generic" (movingWindow) or "particular" (movingMax, etc). we have to do it through a separate parameter.  Let me push up what I mean... stay tuned.

**Comment by shanson7** [comment]
> I see. the difference is that `aggregate()` doesn't set a default `func` whereas `movingWindow` does (see https://graphite.readthedocs.io/en/latest/functions.html)
> so we can't use the `fn` parameter to differentiate whether the invocation was "generic" (movingWindow) or "particular" (movingMax, etc). we have to do it through a separate parameter. Let me push up what I mean... stay tuned.

Couldn't we still though? We could move where the default `fn` is set from the constructor to `Exec`

**Comment by Dieterbe** [comment]
Signature: we can't specify default values for arguments, but we could update the .fn property there, but that would cause mayhem in case Signature() is called again (unlikely with the current code, but that shouldn't be a concern)
Exec: yes, we could apply a default here if it was unset

neither of these approaches seems better or cleaner that what i'm proposing, so let me push what i'm working on in a few minutes. stay tuned.

**Pull request closed without merging**
